lib/unix/git_unix.ml

```diff
@@ -36,44 +36,150 @@ module M = struct
-      Lwt.finalize
-        (fun () -> fn (p#stdout, p#stdin))
-        (fun () -> let _ = p#close in return_unit)
-    | Some "git" ->
-      Log.debug "Connecting to %s" (Uri.to_string uri);
-      let resolver = Resolver_lwt_unix.system in
-      Resolver_lwt.resolve_uri ~uri resolver >>= fun endp ->
-      let ctx = Conduit_lwt_unix.default_ctx in
-      Conduit_lwt_unix.endp_to_client ~ctx endp >>= fun client ->
-      Conduit_lwt_unix.connect ~ctx client >>= fun (_flow, ic, oc) ->
-      Lwt.finalize
-        (fun () ->
-           begin match init with
-             | None   -> return_unit
-             | Some s -> write oc s
-           end >>= fun () ->
-           fn (ic, oc))
-        (fun ()  -> Lwt_io.close ic)
-   | Some x ->
-      (* XXX: make it work for smart-HTTP *)
-      (* XXX: make it work over SSL *)
+      | Some x -> x
+    in
+    let user = match Uri.userinfo uri with
+      | None   -> ""
+      | Some u -> u ^ "@"
+    in
+    let cmd = match init with
+      | None   -> [| "ssh"; user ^ host; |]
+      | Some x -> [| "ssh"; user ^ host; x |]
+    in
+    Log.info "Executing '%s'" (String.concat " " (Array.to_list cmd));
+    let env = Unix.environment () in
+    let p = Lwt_process.open_process_full ~env ("ssh", cmd) in
+    Lwt.finalize
+      (fun () -> fn (p#stdout, p#stdin))
+      (fun () -> let _ = p#close in return_unit)
+
+  let with_conduit ?init uri fn =
+    Log.debug "Connecting to %s" (Uri.to_string uri);
+    let resolver = Resolver_lwt_unix.system in
+    Resolver_lwt.resolve_uri ~uri resolver >>= fun endp ->
+    let ctx = Conduit_lwt_unix.default_ctx in
+    Conduit_lwt_unix.endp_to_client ~ctx endp >>= fun client ->
+    Conduit_lwt_unix.connect ~ctx client >>= fun (_flow, ic, oc) ->
+    Lwt.finalize
+      (fun () ->
+         begin match init with
+           | None   -> return_unit
+           | Some s -> write oc s
+         end >>= fun () ->
+         fn (ic, oc))
+      (fun ()  -> Lwt_io.close ic)
+
+  let http_call ?headers meth uri fn =
+    let headers = match headers with None -> Cohttp.Header.init () | Some h -> h in
+    let callback (ic, oc) =
+      let req = match meth with
+        | `GET ->
+          Cohttp.Request.make_for_client ~headers ~chunked:false
+            (meth :> Cohttp.Code.meth) uri
+        | `POST ->
+          Cohttp_lwt_unix.Request.make_for_client ~headers ~chunked:true
+            (meth :> Cohttp.Code.meth) uri
+      in
+      let http_oc =
+        match meth with
+        | `GET  -> oc
+        | `POST ->
+          let writer =
+            Cohttp_lwt_unix.Request.make_body_writer ~flush:true req oc
+          in
+          Lwt_io.make ~mode:Lwt_io.output ~close:(fun () -> Lwt_io.close oc)
+            (fun bytes off len ->
+               let chunk = Bytes.create len in
+               Lwt_bytes.blit_to_bytes bytes off chunk 0 len;
+               Cohttp_lwt_unix.Request.write_body writer chunk >>= fun () ->
+               return len)
```

Comment:
Hm, I wonder if this would be easier with a Cohttp_lwt_cstruct backend instead?  @djs55 wanted this for the xapi socket backend too